MulticastDeserializationTest.test

_master_ (commit 7fdfd902b26738ef7f31bf1de646c445ceffc2e9)

Failed on ZingJDK8 : http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-3.x-ZingJDK8/68/testReport/com.hazelcast.cluster/MulticastDeserializationTest/test/

**Stacktrace**
```
java.lang.AssertionError: Untrusted deserialization is possible
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertFalse(Assert.java:64)
	at com.hazelcast.cluster.MulticastDeserializationTest.test(MulticastDeserializationTest.java:81)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:745)
```

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MulticastService.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MulticastService.java
@@ -48,6 +48,11 @@ import static com.hazelcast.util.EmptyStatement.ignore;
 
 public final class MulticastService implements Runnable {
 
+    /**
+     * IP address of a multicast group. If not set, configuration is read from the {@link MulticastConfig} configuration.
+     */
+    public static final String SYSTEM_PROPERTY_MULTICAST_GROUP = "hazelcast.multicast.group";
+
     private static final int SEND_OUTPUT_SIZE = 1024;
     private static final int DATAGRAM_BUFFER_SIZE = 64 * 1024;
     private static final int SOCKET_BUFFER_SIZE = 64 * 1024;
@@ -131,7 +136,7 @@ public final class MulticastService implements Runnable {
             }
             multicastSocket.setReceiveBufferSize(SOCKET_BUFFER_SIZE);
             multicastSocket.setSendBufferSize(SOCKET_BUFFER_SIZE);
-            String multicastGroup = System.getProperty("hazelcast.multicast.group");
+            String multicastGroup = System.getProperty(SYSTEM_PROPERTY_MULTICAST_GROUP);
             if (multicastGroup == null) {
                 multicastGroup = multicastConfig.getMulticastGroup();
             }

--- a/hazelcast/src/test/java/com/hazelcast/cluster/MulticastDeserializationTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/cluster/MulticastDeserializationTest.java
@@ -30,7 +30,9 @@ import java.net.InetAddress;
 import java.net.MulticastSocket;
 import java.nio.ByteBuffer;
 
-import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.Rule;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
@@ -40,9 +42,11 @@ import com.hazelcast.config.JavaSerializationFilterConfig;
 import com.hazelcast.config.JoinConfig;
 import com.hazelcast.core.Hazelcast;
 import com.hazelcast.instance.impl.HazelcastInstanceFactory;
+import com.hazelcast.internal.cluster.impl.MulticastService;
 import com.hazelcast.internal.serialization.impl.SerializationConstants;
 import com.hazelcast.nio.Packet;
 import com.hazelcast.test.HazelcastSerialClassRunner;
+import com.hazelcast.test.OverridePropertyRule;
 import com.hazelcast.test.annotation.QuickTest;
 
 import example.serialization.TestDeserialized;
@@ -59,8 +63,17 @@ public class MulticastDeserializationTest {
     // TTL = 0 : restricted to the same host, won't be output by any interface
     private static final int MULTICAST_TTL = 0;
 
-    @After
-    public void tearDown() {
+    @Rule
+    public OverridePropertyRule multicastGroupOverride = OverridePropertyRule
+            .clear(MulticastService.SYSTEM_PROPERTY_MULTICAST_GROUP);
+
+    @Before
+    public void before() {
+        cleanup();
+    }
+
+    @AfterClass
+    public static void cleanup() {
         TestDeserialized.isDeserialized = false;
         HazelcastInstanceFactory.terminateAll();
     }

--- a/hazelcast/src/test/java/com/hazelcast/instance/MulticastLoopbackModeTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/instance/MulticastLoopbackModeTest.java
@@ -21,11 +21,14 @@ import com.hazelcast.config.MulticastConfig;
 import com.hazelcast.cluster.Cluster;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.instance.impl.HazelcastInstanceFactory;
+import com.hazelcast.internal.cluster.impl.MulticastService;
 import com.hazelcast.test.HazelcastSerialClassRunner;
 import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.OverridePropertyRule;
 import com.hazelcast.test.annotation.QuickTest;
 import org.junit.After;
 import org.junit.Before;
+import org.junit.Rule;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
@@ -49,25 +52,22 @@ import static org.junit.Assume.assumeFalse;
 @Category(QuickTest.class)
 public class MulticastLoopbackModeTest extends HazelcastTestSupport {
 
-    private String multicastGroup;
     private HazelcastInstance hz1;
     private HazelcastInstance hz2;
+
+    @Rule
+    public OverridePropertyRule multicastGroupOverride = OverridePropertyRule
+            .clear(MulticastService.SYSTEM_PROPERTY_MULTICAST_GROUP);
 
     @Before
     public void setUpTests() {
         assumeFalse("This test can be processed only if your host has no configured network interface.",
                 hasConfiguredNetworkInterface());
-        multicastGroup = System.clearProperty("hazelcast.multicast.group");
     }
 
     @After
     public void tearDownTests() {
         HazelcastInstanceFactory.terminateAll();
-        if (multicastGroup == null) {
-            System.clearProperty("hazelcast.multicast.group");
-        } else {
-            System.setProperty("hazelcast.multicast.group", multicastGroup);
-        }
     }
 
     private void createTestEnvironment(boolean loopbackMode) throws Exception {

